tinygrad/tensor.py

```diff
@@ -314,21 +316,50 @@ def num_zeros(step, dim_sz): return 0 if step == 1 or (y := dim_sz % step) == 0
       final_slice = reduce(operator.add, (((0, sh), (0, 1)) for sh in new_shape), ())
       sliced_tensor = reshaped_tensor.shrink(final_slice)
     final_shape = []
+    sub = [0] * len(tensor_found)
     it_shape = iter(new_shape)
-    for i in orig_slices:
+    for n,i in enumerate(orig_slices):
       if isinstance(i, (int, slice)):
         dim_shape = next(it_shape)
         if isinstance(i, slice): final_shape.append(dim_shape)
+        elif tensor_found:
+          for n_ in range(len(tensor_found)):
+            if tensor_found[n_][0] > n: sub[n_] -= 1
       else: # i is None
         final_shape.append(1)
-    return sliced_tensor.reshape(tuple(final_shape))  # Reshape
-
-  def gather(self, idx, dim):
-    idx = (idx < 0).where(idx+self.shape[dim], idx) # Turn neg idx pos
-    new_self = self.reshape(*self.shape[:dim+1], *[1]*idx.ndim, *self.shape[dim+1:])
-    arange = Tensor.arange(self.shape[dim], dtype=dtypes.int32, requires_grad=False).reshape(*[1]*dim, self.shape[dim], *[1]*idx.ndim, *[1]*(self.ndim-dim-1))
-    new_idx = idx.reshape(*[1]*dim, 1, *idx.shape, *[1]*(self.ndim-dim-1))
-    return (new_self * (arange == new_idx)).sum(dim)
+    ret = sliced_tensor.reshape(tuple(final_shape))  # Reshape
+    # Fancy/tensor indexing
+    if tensor_found:
+      for i,s in enumerate(sub): tensor_found[i] = (tensor_found[i][0]+s, tensor_found[i][1])
+      dim = [i[0] for i in tensor_found]
+      idx = [(i[1] < 0).where(i[1]+ret.shape[i[0]], i[1]) for i in tensor_found]
```

Comment:
is it possible to get rid of the comparison and where here? can we do a multiplication and add?

Reply:
like this? `idx = [-i[1].sign().minimum(0) * ret.shape[i[0]] + i[1]  for i in tensor_found]` 🤣  🤣  🤣 
I never know if I'm doing something stupid